### PR TITLE
Added kl_div loss to ivy experimental api

### LINKED ISSUE #22996: kl_div

### PR DESCRIPTION
**Pull request by Vismay-dev**
# PR Description 

This PR has added support for kl_div loss to the ivy experimental api. It has complete support for kl_div in the backends, including ivy array and container instance methods, and test function.

Close #22996

## Checklist 

- [x] Did you add a function?
- [x] Did you add the tests?
- [x] Did you follow the steps we provided?

### CODE REVIEW

**Comment by Vismay-dev** [comment]
Hey @mobley-trent! Could you please review this pull request?

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 1]
```diff
@@ -55,6 +55,54 @@ def test_huber_loss(
     )
 
 
+# kl_div
+@handle_test(
+    fn_tree="functional.ivy.experimental.kl_div",
+    dtype_and_input=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
```
Comment (mobley-trent) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:62]: When using the decorator for supported / unsupported dtypes, always make sure to use valid dtypes during testing

```suggestion
        available_dtypes=helpers.get_dtypes("valid"),
```

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 2]
```diff
@@ -55,6 +55,54 @@ def test_huber_loss(
...
+        min_value=1e-07,
+        max_value=1,
+        allow_inf=False,
+        min_num_dims=1,
+        max_num_dims=3,
+        min_dim_size=3,
+    ),
+    dtype_and_target=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("float"),
```
Comment (mobley-trent) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:71]:
```suggestion
        available_dtypes=helpers.get_dtypes("valid"),
```

File: ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py [hunk 3]
```diff
@@ -55,6 +55,54 @@ def test_huber_loss(
...
+        test_flags=test_flags,
+        backend_to_test=backend_fw,
+        fn_name=fn_name,
+        on_device=on_device,
+        input=input[0],
+        target=target[0],
+        atol_=1e-02,
+        reduction=reduction,
+    )
+
```
Comment (mobley-trent) [inline ivy_tests/test_ivy/test_functional/test_experimental/test_nn/test_losses.py:104]: Even after making those changes, tests still fail due to some inconsistencies between the dtypes of the return of the ground truth and your function.

**Review by mobley-trent — CHANGES_REQUESTED** [review]

**Comment by mobley-trent** [comment]
Hi @Vismay-dev your PR looks awesome ✨ 
Once the tests are up and running ok feel free to ping me 🙂

**Comment by Vismay-dev** [comment]
@mobley-trent there are no new failures introduced by the latest commit of test_kl_div. please review and guide me.

**Comment by mobley-trent** [comment]
Can you figure out what is causing the returned dtypes to be different ? Try testing the ivy function you've implemented along with its native version with the failing examples.

**Comment by Vismay-dev** [comment]
@mobley-trent from the running the native and backend-implemented functions separately, I couldn't notice any discrepancy, except numpy's (default backend) implicit conversions to float64. 

I pulled all recent changes in the base branch into my fork and that **seemed to solve the problem with different returned dtypes.** 

The current errors are to do with numpy incompatibility with bfloat16 (which is external to this PR), **which has made other previously implemented loss functions fail as well.** There aren't any other errors stemming from kl_div. My local code-space tests are returning 100% passed.

Please review.

File: ivy/functional/ivy/experimental/losses.py [hunk 4]
```diff
@@ -408,3 +408,76 @@ def soft_margin_loss(
...
+    if reduction == "sum":
+        loss = ivy.sum(loss, out=out)
+    elif reduction == "mean":
+        loss = ivy.mean(loss, out=out)
+    elif reduction == "batchmean":
+        loss = ivy.sum(loss, out=out) / size[0]
+    else:
+        loss = ivy.inplace_update(out, loss) if out is not None else loss
+
+    return loss
```
Comment (mobley-trent) [inline ivy/functional/ivy/experimental/losses.py:483]:
```suggestion

    return ivy.inplace_update(out, loss) if out is not None else loss
```

File: ivy/functional/ivy/experimental/losses.py [hunk 5]
```diff
@@ -408,3 +408,76 @@ def soft_margin_loss(
...
+    ivy.array(0.0916)
+
+    >>> input = ivy.array([0.2, 0.8], [0.5, 0.5])
+    >>> target = ivy.array([0.6, 0.4], [0.3, 0.7])
+    >>> ivy.kl_div(input, target, reduction='none')
+    ivy.array([0.0378], [0.1453])
+    """
+    size = ivy.shape(input)
+    # if len(size) < 1:
+    #     size = [1]
```
Comment (mobley-trent) [inline ivy/functional/ivy/experimental/losses.py:470]: Remove this if it is not needed

File: ivy/functional/backends/numpy/experimental/losses.py [hunk 6]
```diff
@@ -70,3 +70,29 @@ def soft_margin_loss(
...
+    loss = np.sum(input * np.log(input / target), axis=-1)
+
+    if reduction == "mean":
+        loss = np.mean(loss)
+    elif reduction == "sum":
+        loss = np.sum(loss)
+    elif reduction == "batchmean":
+        loss = np.divide(np.sum(loss), size[0])
+    else:
+        pass
```
Comment (mobley-trent) [inline ivy/functional/backends/numpy/experimental/losses.py:97]:
```suggestion
```

File: ivy/functional/backends/tensorflow/experimental/losses.py [hunk 7]
```diff
@@ -62,3 +62,29 @@ def soft_margin_loss(
...
+    #     size = [1]
+
+    if reduction == "mean":
+        loss = tf.math.reduce_mean(loss)
+    elif reduction == "sum":
+        loss = tf.math.reduce_sum(loss)
+    elif reduction == "batchmean":
+        loss = tf.math.reduce_sum(loss) / tf.cast(size[0], dtype=tf.float32)
+    else:
+        pass
```
Comment (mobley-trent) [inline ivy/functional/backends/tensorflow/experimental/losses.py:88]:
```suggestion
```

I don't think this logic is necessary

File: ivy/functional/backends/jax/experimental/losses.py [hunk 8]
```diff
@@ -56,3 +56,26 @@ def soft_margin_loss(
...
+
+    loss = jnp.sum(input * jnp.log(input / target), axis=-1)
+    if reduction == "mean":
+        loss = jnp.mean(loss)
+    elif reduction == "sum":
+        loss = jnp.sum(loss)
+    elif reduction == "batchmean":
+        loss = jnp.divide(jnp.sum(loss), size[0])
+    else:
+        pass
```
Comment (mobley-trent) [inline ivy/functional/backends/jax/experimental/losses.py:80]:
```suggestion
```

File: ivy/functional/backends/jax/experimental/losses.py [hunk 9]
```diff
@@ -56,3 +56,26 @@ def soft_margin_loss(
...
+def kl_div(
+    input: JaxArray,
+    target: JaxArray,
+    /,
+    *,
+    reduction: Optional[str] = "mean",
+) -> JaxArray:
+    size = jnp.shape(input)
+    # if len(size) < 1:
+    #     size = [1]
```
Comment (mobley-trent) [inline ivy/functional/backends/jax/experimental/losses.py:70]:
```suggestion
```

File: ivy/functional/backends/tensorflow/experimental/losses.py [hunk 10]
```diff
@@ -62,3 +62,29 @@ def soft_margin_loss(
...
+    /,
+    *,
+    reduction: Optional[str] = "mean",
+) -> tf.Tensor:
+    loss = tf.reduce_sum(input * tf.math.log(input / target), axis=-1)
+
+    size = tf.shape(input)
+    # if len(size) < 1:
+    #     size = [1]
+
```
Comment (mobley-trent) [inline ivy/functional/backends/tensorflow/experimental/losses.py:80]:
```suggestion
```

**Comment by Vismay-dev** [comment]
@mobley-trent i've made the changes 👍.

**Comment by mobley-trent** [comment]
Could you also remove the changes to `demos` please 🙂

**Comment by Vismay-dev** [comment]
@mobley-trent I've removed the changes to demos.

**Review by mobley-trent — APPROVED** [review]
Just curious how did you manage to remove the changes  😅 ?
I think there was a force push to `demos`

**Comment by Vismay-dev** [comment]
I overwrote the current branch folder with the origin/main version of docs/demos. That reset any changes within docs/demos to fit the most recently pulled version of the base branch which was upstream in my fork 🙂.

Honestly, I don't quite understand why there were changes to docs/demos or what prompted the force push in the first place. My understanding is that it was because of resolving a merge conflict.
